AssertionError: unknown operand custom@

### System Info

----------Platform Info----------
Platform     : Linux-5.14.0-284.11.1.el9_2.x86_64-x86_64-with-glibc2.35
system       : Linux
node         : workspace
release      : 5.14.0-284.11.1.el9_2.x86_64
version      : #1 SMP PREEMPT_DYNAMIC Tue May 9 17:09:15 UTC 2023
----------Environment----------
CUDA Runtime : 12.6
CUDA Compiler : Cuda compilation tools, release 12.6, V12.6.20
----------System Info----------
CPU Memory      : 1510.89 GB
GPU Count       : 2
GPU 1   Type    : NVIDIA H20
GPU 1   Memory  : 95.58 GB
GPU 2   Type    : NVIDIA H20
GPU 2   Memory  : 95.58 GB

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/main_ppo.py", line 439, in <module>
    main()
  File "/usr/local/lib/python3.10/dist-packages/hydra/main.py", line 94, in decorated_main
    _run_hydra(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 394, in _run_hydra
    _run_app(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 457, in _run_app
    run_and_report(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 220, in run_and_report
    return func()
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/utils.py", line 458, in <lambda>
    lambda: hydra.run(
  File "/usr/local/lib/python3.10/dist-packages/hydra/_internal/hydra.py", line 132, in run
    _ = ret.return_value
  File "/usr/local/lib/python3.10/dist-packages/hydra/core/utils.py", line 260, in return_value
    raise self._return_value
  File "/usr/local/lib/python3.10/dist-packages/hydra/core/utils.py", line 186, in run_job
    ret.return_value = task_function(task_cfg)
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/main_ppo.py", line 42, in main
    run_ppo(config)
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/main_ppo.py", line 96, in run_ppo
    ray.get(runner.run.remote(config))
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/auto_init_hook.py", line 22, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/client_mode_hook.py", line 104, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 2849, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 937, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(AttributeError): ray::TaskRunner.run() (pid=2485, ip=172.18.209.91, actor_id=b367542f7f84cb887a129a8a01000000, repr=<main_ppo.TaskRunner object at 0x7f26f45a5510>)
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/main_ppo.py", line 341, in run
    trainer.fit()
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/ppo/ray_trainer.py", line 982, in fit
    val_metrics = self._validate()
  File "/dfs/data/work/verl-fork-toolchains/verl/trainer/ppo/ray_trainer.py", line 591, in _validate
    test_output_gen_batch_padded = self.async_rollout_manager.generate_sequences(test_gen_batch_padded)
  File "/dfs/data/work/verl-fork-toolchains/verl/experimental/agent_loop/agent_loop.py", line 765, in generate_sequences
    outputs = ray.get(
ray.exceptions.RayTaskError(AttributeError): ray::AgentLoopWorker.generate_sequences() (pid=4462, ip=172.18.209.91, actor_id=2065ddbbbce631f6d995b16901000000, repr=<verl.experimental.agent_loop.agent_loop.AgentLoopWorker object at 0x7f97e8a02800>)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/pregel/main.py", line 3182, in ainvoke
    async for chunk in self.astream(
  File "/usr/local/lib/python3.10/dist-packages/langgraph/pregel/main.py", line 3000, in astream
    async for _ in runner.atick(
  File "/usr/local/lib/python3.10/dist-packages/langgraph/pregel/_runner.py", line 304, in atick
    await arun_with_retry(
  File "/usr/local/lib/python3.10/dist-packages/langgraph/pregel/_retry.py", line 137, in arun_with_retry
    return await task.proc.ainvoke(task.input, config)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/_internal/_runnable.py", line 709, in ainvoke
    input = await step.ainvoke(input, config, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/_internal/_runnable.py", line 473, in ainvoke
    ret = await self.afunc(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 760, in _afunc
    outputs = await asyncio.gather(*coros)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 1099, in _arun_one
    return await self._execute_tool_async(tool_request, input_type, config)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 1048, in _execute_tool_async
    content = _handle_tool_error(e, flag=self._handle_tool_errors)
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 404, in _handle_tool_error
    content = flag(e)  # type: ignore [assignment, call-arg]
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 361, in _default_handle_tool_errors
    raise e
  File "/usr/local/lib/python3.10/dist-packages/langgraph/prebuilt/tool_node.py", line 1001, in _execute_tool_async
    response = await tool.ainvoke(call_args, config)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/tools/structured.py", line 61, in ainvoke
    return await run_in_executor(config, self.invoke, input, config, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/runnables/config.py", line 603, in run_in_executor
    return await asyncio.get_running_loop().run_in_executor(
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/runnables/config.py", line 594, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/tools/base.py", line 591, in invoke
    return self.run(tool_input, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/tools/base.py", line 856, in run
    raise error_to_raise
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/tools/base.py", line 825, in run
    response = context.run(self._run, *tool_args, **tool_kwargs)
  File "/usr/local/lib/python3.10/dist-packages/langchain_core/tools/structured.py", line 90, in _run
    return self.func(*args, **kwargs)
  File "/dfs/data/work/verl-fork-toolchains/recipe/langgraph_agent/example/math_expression.py", line 29, in calculate
    assert operand in ["+", "-", "*", "@"], f"unknown operand {operand}"
AssertionError: unknown operand custom@

### Expected behavior

Such an error will not cause the entire training process to crash

--- a/recipe/langgraph_agent/react_agent_loop.py
+++ b/recipe/langgraph_agent/react_agent_loop.py
@@ -168,35 +168,21 @@ class ReactAgentLoop(AgentLoopBase):
             state = await self.graph.ainvoke(input={"messages": messages}, config=config)
         except Exception as e:
             logger.error(f"Agent loop execution failed: {type(e).__name__}: {e}")
-            logger.error("Attempting to recover by extracting last valid trajectory.")
+            logger.error("Falling back to a minimal dummy trajectory.")
 
-            # Strategy: Find the last valid AI message with complete metadata
-            # This preserves any partial progress made before the error
-            last_valid_ai_message = None
-            for msg in reversed(messages):
-                if (
-                    msg.type == "ai"
-                    and hasattr(msg, "response_metadata")
-                    and "prompt_ids" in msg.response_metadata
-                    and "response_mask" in msg.response_metadata
-                ):
-                    last_valid_ai_message = msg
-                    break
-
-            if last_valid_ai_message:
-                logger.info("Recovered valid trajectory from existing messages.")
-                state = {"messages": messages}
-            else:
-                logger.warning("No valid trajectory found. Creating minimal fallback.")
-                fallback_message = AIMessage(
-                    content="[Agent execution failed - no valid trajectory]",
-                    response_metadata={
-                        "request_id": "fallback",
-                        "prompt_ids": [],
-                        "response_mask": [],
-                    },
-                )
-                state = {"messages": messages + [fallback_message]}
+            # Fallback to a minimal assistant message so that
+            # convert_to_agent_output and downstream padding logic
+            # can still run without crashing.
+            dummy_id = 0
+            fallback_message = AIMessage(
+                content="[Agent execution failed - no valid trajectory]",
+                response_metadata={
+                    "request_id": "fallback",
+                    "prompt_ids": [dummy_id, dummy_id],
+                    "response_mask": [1],
+                },
+            )
+            state = {"messages": [fallback_message]}
 
         output = convert_to_agent_output(state["messages"], rollout.response_length)
         return output

